File: server/server/dtos.py
from dataclasses import dataclass
from dataclasses_json import DataClassJsonMixin

EVENT_HEARTBEAT = "ping"

EVENT_CAPACITY_REQUEST = "capacity_request"
EVENT_CAPACITY_RESPONSE = "capacity_response"

ROOM_CAPACITY_BROADCAST = "capacity"

EVENT_CANCEL_REQUEST = "cancel_request"
EVENT_CANCEL_RESPONSE = "cancel_response"

EVENT_MESSAGE_REQUEST = "message_request"
EVENT_MESSAGE_RESPONSE = "message_response"

EVENT_SERVER_ERROR = "server_error"

@dataclass
class CapacityRequest(DataClassJsonMixin):
    challenge_id: int

@dataclass
class CapacityResponse(DataClassJsonMixin):
    queue_size: int
    diff: int
    ready: bool

@dataclass
class MessageRequest(DataClassJsonMixin):
    challenge_id: int
    content: str

@dataclass
class MessageResponse(DataClassJsonMixin):
    content: str
    clawback: bool
    stop: bool

@dataclass
class ServerErrorResponse(DataClassJsonMixin):
    error_msg: str

File: clientapp/src/app/dtos.ts
//Request capacity
export const EVENT_CAPACITY_REQUEST="capacity_request";
export interface CapacityRequest{
    challenge_id: number;
}

export const EVENT_HEARTBEAT="ping";

export const EVENT_CAPACITY_RESPONSE="capacity_response";
export interface CapacityResponse{
    queue_size: number; //Size of the queue initially
    diff: number; //Difference to apply to the size of the queue
    ready: boolean; //Indicates that the user can request
}

//Cancel capacity request or message request
export const EVENT_CANCEL_REQUEST="cancel_request";

//Server confirms that the request was cancelled properly.
export const EVENT_CANCEL_RESPONSE="cancel_response";

//Once on position 0 for capacityResponse client has 5s to send a message request
//before being assigned to another client.
export const EVENT_MESSAGE_REQUEST="message_request";
export interface MessageRequest{
    challenge_id: number;
    content: string;
}

//Output from the LLM to the client
export const EVENT_MESSAGE_RESPONSE="message_response";
export interface MessageResponse{
    content: string;
    clawback: boolean; //Indicates to the client to remove all the previous text
    stop: boolean; //Indicates that this is the final message
}

//Ways for the server to send errors to the client
export const EVENT_SERVER_ERROR="server_error";
export interface ServerErrorResponse{
    error_msg: string;
}